src/react-hooks-nesting-walker/error-messages.ts

```diff
@@ -17,4 +17,6 @@ export const ERROR_MESSAGES = {
     'A hook cannot be used inside of another function',
   invalidFunctionExpression: 'A hook cannot be used inside of another function',
```

Comment:
Actually, I've had a thought - wouldn't this PR be as simple as just modifying this message to make it more complete?

I didn't investigate it deeply, but this looks like the error message that is reported exactly in the situations we're trying to improve. Let me know if I'm mistaken

Reply:
That is a good point. But I don't think so. There is one illegal scenario where a hook is instantiated within an anonymous function that within an anonymous legal component function. For that we need the "A hook cannot be used inside of another function" error message.

I added one test scenario to cover for it.

Reply:
Ok, I think I know what you mean now :smile: An equivalent of [`isSomewhereInsideComponentOrHook` from the ESLint rule](https://github.com/facebook/react/blob/51947a14bb24bd151f76f6fc0acdbbc404de13f7/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L500)

To implement that, we'd probably need to keep track of the surrounding context (all ancestor functions for a given hook call). For example, for:

```ts
const Functor = function() {
  const cb = useCallback(() => {
    const r = React.useRef()
  }, [])
}
```

for the innermost `useRef`, the function stack is:

1. The anonymous function expression assigned to `Functor`
2. The arrow function passed to `useCallback`

The top of the stack would tell us if it's a valid hook call or not. If we wanted to get the `isSomewhereInsideComponentOrHook`, we'd need to check if there's a valid component/hook declaration somewhere on that stack. In this case, there is (1), so we could show the `invalidFunctionExpression` warning, and if there wasn't, we'd show `anonymousFunctionIllegalCallback`.

Not sure if that's robust enough to handle all cases :thinking: